### DISCUSSION TITLE
how to customize post method

### DISCUSSION THREAD

**Discussion by hyh1618** [thomaxxl/safrs#126] [General]
Can anyone please tell me how do I add actions into the generated POST action? Post will create a record into class object table, but I need do more after that. Thanks a lot!

**Comment by thomaxxl** [comment]
You can do this by hooking HTTP POST, https://github.com/thomaxxl/safrs/wiki/HTTP-Methods . Hooking relationship post is a bit more complicated.

There are also other ways to achieve what you want, using sqlalchemy triggers for example.

**Reply by hyh1618** [reply]
Thanks a lot thomaxxl. what i need is to send the generated new record/class object to somewhere, either hooking or trigger will work, can you please point me to more details about it? as I'm new here. I tried example hooking as below but it throw error " ERROR: _s_post() missing 1 required positional argument: 'self' "
    def _s_post(self, *args, **kwargs):
        print(kwargs)
I appreciate much.

**Reply by thomaxxl** [reply]
Hi,

I am on holiday atm without computer so I can't check this myself. Another approach is used here, where decorators are used to hook HTTP requests.

https://github.com/thomaxxl/safrs/blob/master/examples/authentication/demo_post_auth.py

**Reply by hyh1618** [reply]
Thanks, please help when you can. I tried that decorator, seems its before the record creation, I can't get data.

**Comment by thomaxxl** [comment]
This is one approach (using flask [request](https://flask.palletsprojects.com/en/2.2.x/api/#incoming-request-data) and [response](https://flask.palletsprojects.com/en/2.2.x/api/#flask.Response))
https://gist.github.com/thomaxxl/7f009114b0c86b7205611a60a0704abf

can you try it please?

**Comment by thomaxxl** [comment]
For the record, `_s_post` is a bit more complicated than described in the wiki as it's a `classmethod`.
[This should work](https://gist.github.com/thomaxxl/48fbd9c0401e677a588c8048522310bd)

```python
    @classmethod
    def _s_post(cls, *args, **kwargs):
        print(kwargs)
        result = User(*args, **kwargs)
        print(result)
        return result
``` 

I will update the documentation.

**Comment by hyh1618** [comment]
Thanks Thomaxxl/Safrs,

At this point (in your code) the new record already committed in DB? I need
call another function from here which will need this record already in DB.

Thanks!

**Comment by thomaxxl** [comment]
Records are automatically committed when created in `SAFRSBase.__init__`, unless `SAFRSBase._s_auto_commit` is `False` (default is `True` for now but this might change in future versions).

**Comment by hyh1618** [comment]
Hi Thomaxxl/Safrs,

Thanks. Yes it committed and can be queried, but it throw exception and
exit at end:
Error: 'NoneType' object has no attribute '_s_object_id'

I'm on Python 3.9 and latest safrs

**Reply by thomaxxl** [reply]
I think you forgot to return the created class ( ` return result `)

see https://gist.github.com/thomaxxl/48fbd9c0401e677a588c8048522310bd

**Comment by hyh1618** [comment]
Thanks Thomas, it is working now.
Can you please show me an example to display description info for a column?
It seems not works for 2.12.3. the table description works. Thanks a lot
for your help.

**Reply by thomaxxl** [reply]
You want to change the description shown in the swagger?

You can modify the spec by providing the `custom_swagger` argument to `SAFRSAPI`: https://github.com/thomaxxl/safrs/blob/b411752adb9cc7af2828fa4ccefe89fb6cd469b4/examples/mini_examples/ex12_swagger.py#L53

**Comment by hyh1618** [comment]
Yes. I mean to customize the default service, such as get or post, to add
description such as "use yyy-mm-dd" as format for field created, like that.
Thanks!

**Reply by thomaxxl** [reply]
The only way to do this is by modifying the swagger manually. You can import and edit the spec with https://editor.swagger.io/ , then host the spec and swagger ui elsewhere, or merge using the above `custom_swagger` argument.

**Comment by hyh1618** [comment]
Thanks, is it possible to send a example of custom_swagger argument for
service level description?

**Comment by hyh1618** [comment]
Hi Thomas,

On another topic, I see your demo_auth.py, run it, the page does not pop up
to ask userid/pw, how do you suggest to let user type in userid & password
in UI? Thanks a lot.

**Reply by thomaxxl** [reply]
Hi,

Can you next time open a new discussion thread for new questions please?

API authentication is usually implemented using a token passed in the [HTTP Authorization header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization).
You should add this header when you execute HTTP requests in the frontend, so this depends a bit on which frontend you use.

Regarding the swagger, this will require some work if you want to customize it. The `custom_swagger` argument will merge the given items with the existing swagger. So, you'll have to find out which items you want to modify, then pass this as a dictionary to `custom_swagger`.

**Comment by hyh1618** [comment]
OK, I will.
Yes I know to add an item into custom_swagger dict, but I don't know the
naming pattern for a service endpoint, is it classname_fieldname? any doc
link on this? Thanks.
